### ISSUE TITLE
ShadowDOM and chromedp.ByJSPath

### ISSUE THREAD

**Issue by A139834** [chromedp/chromedp#376]
Hi,
I have been trying to access elements within a ShadowDOM. Puppeteer has solved this problem via JS Path evaluation: https://github.com/GoogleChrome/puppeteer/issues/858#issuecomment-438540596

I have coded up a sample to add `chromedp.ByJSPath` with this api:

```go
chromedp.Click(`document.querySelector('#container').shadowRoot.querySelector('#foo')`, chromedp.ByJSPath),
```

Here is my quick implementation:

```go
// ByJSPath is a query option to select elements by JS Path
// this allows selection of nodes within ShadowDOM nodes
func ByJSPath(s *Selector) {
	ByFunc(func(ctx context.Context, n *cdp.Node) ([]cdp.NodeID, error) {
		// set awaitPromise
		p := runtime.Evaluate(s.selAsString()).WithAwaitPromise(true)

		// evaluate
		v, exp, err := p.Do(ctx)
		if err != nil {
			return nil, err
		}
		if exp != nil {
			return nil, exp
		}

		// use the ObjectID from the evaluation to get the nodeID
		nodeID, err := dom.RequestNode(v.ObjectID).Do(ctx)
		if err != nil {
			return nil, err
		}

		if nodeID == cdp.EmptyNodeID {
			return []cdp.NodeID{}, nil
		}

		return []cdp.NodeID{nodeID}, nil
	})(s)
}
```

This works and appears to be quick. Maybe you have a better solution? Or happy to do a PR.

**Comment by seddonm1** [comment]
Hi,
If anyone is interested I have created proof-of-concept for using JSPath instead of XPath to allow ByJSPath (above) to behave like any other selector. JSPath allows full traversal of the DOM including any ShadowDOM.

So adding `ByJSPath` and refactoring the included Javascript functions in:
https://github.com/seddonm1/chromedp

And adding `jspath` vs `xpath` in:
https://github.com/seddonm1/cdproto

You can do track changes to see diffs.

**Comment by kenshaw** [comment]
@seddonm1 please submit a PR for this.

**Comment by seddonm1** [comment]
@kenshaw I can but this is a relatively large change given the reliance on XPATH (which does not support ShadowDOM)  for a lot of the Selector functions. This would need agreement from @mvdan and others to switch to JSPATH.

**Comment by kenshaw** [comment]
This would not become the default option -- adding an additional selector type is more than fine. Dan can of course weigh in, but I don't see any reason to not add this. The only thing missing from the above is a unit test covering the selector.

**Comment by seddonm1** [comment]
Hi, happy to submit the PR but it will not achieve much given that most chromedp Actions assume they will be given an XPATH: https://github.com/chromedp/chromedp/blob/master/js.go#L70 (which does not work through shadowDom boundaries.

I had to change the cdproto to add a JSPath javascript command generator: https://github.com/seddonm1/cdproto/blob/master/cdp/types.go#L393 

and modify the JS to accept either the XPATH or JSPath command:
https://github.com/seddonm1/chromedp/blob/master/js.go#L70

I guess the big question is if chromedp wants to support JSPath. If so then it needs to be applied everywhere (instead of XPath by default)

**Comment by kenshaw** [comment]
@seddonm1 apologies, I did not realize you weren't the original issue author. I did not mean to cause confusion. Also, note that the default `By` func is `BySearch`, which takes either a XPath or a CSS query. The underlying call is Chrome's [`DOM.performSearch`](https://chromedevtools.github.io/devtools-protocol/tot/DOM#method-performSearch).

@A139834 I'd be more than happy to put in this `By*` func. To simplify things, I'll just add this and a unit test for it. Feel free to submit a PR with this, if you'd like your name/email in the commit history.
